File: 2024/day7.py
f = open('inputs/input_7', 'r')


def is_valid(result, value, numbers, curr):
    if result == value and curr == len(numbers):
        return True
    elif curr == len(numbers):
        return False

    return is_valid(result + numbers[curr], value, numbers, curr+1) or is_valid(result * numbers[curr], value, numbers, curr+1) or is_valid(int(str(result) + str(numbers[curr])), value, numbers, curr+1)


count = 0

for line in f.readlines():
    value, equation = line.split(':')
    value = int(value)

    numbers = list(map(lambda x: int(x), equation.split()))

    if (is_valid(numbers[0], value, numbers, 1)):
        count += value

print(count)

File: 2025/day7.exs
defmodule Day7 do
  def main do
    {_, result} =
      File.stream!("inputs/input_7.txt")
      |> Enum.map(fn line -> String.split(line, "", trim: true) end)
      |> Enum.reduce({[], 0}, fn line, {prev, result} ->
        {indexs, count} =
          Enum.reduce(Enum.with_index(line), {[], 0}, fn {c, i}, {acc, count} ->
            if i not in prev do
              if c == "S", do: {[i | acc], count}, else: {acc, count}
            else
              if c == "^", do: {[i + 1, i - 1 | acc], count + 1}, else: {[i | acc], count}
            end
          end)

        {indexs, result + count}
      end)

    IO.puts(result)
  end
end

Day7.main()
